drivers/at24cxxx/at24cxxx.c

```diff
@@ -0,0 +1,248 @@
+{
+    int check;
+    uint8_t polls = AT24CXXX_MAX_POLLS;
+
+    while ((check =
+                i2c_read_regs(dev->params.i2c, dev->params.dev_addr, pos, data,
+                              len,
+                              I2C_REG16)) == -ENXIO && --polls) {
+        xtimer_usleep(AT24CXXX_POLL_DELAY_MS * US_PER_MS);
+    }
+    DEBUG("_read::i2c_read_regs check: %d; polls: %d\n", check, polls);
+    return check < 0 ? check : (int)len;
+}
+
+static int _write(const at24cxxx_t *dev, uint32_t pos, const uint8_t *data,
+                  size_t len)
+{
+    int check = 0;
+    size_t wr = 0;
+    size_t chunk_len;
+
+    while (!check && (chunk_len = len - wr)) {
+        const uint8_t *chunk_data = data + wr;
+        size_t max_chunk_len =
+            (AT24CXXX_PAGE_SIZE - (pos % AT24CXXX_PAGE_SIZE));
+        chunk_len = chunk_len < max_chunk_len ? chunk_len : max_chunk_len;
+
+        uint8_t polls = AT24CXXX_MAX_POLLS;
+        while ((check =
+                    i2c_write_regs(dev->params.i2c, dev->params.dev_addr, pos,
+                                   chunk_data,
+                                   chunk_len,
+                                   I2C_REG16)) == -ENXIO && --polls) {
+            xtimer_usleep(AT24CXXX_POLL_DELAY_MS * US_PER_MS);
+        }
+        DEBUG("_write::i2c_write_regs check: %d; polls: %d\n", check, polls);
+        if (!check) {
+            wr += chunk_len;
+            pos += chunk_len;
+        }
+    }
+    return wr > 0 ? (int)wr : check;
+}
+
+static int _set(const at24cxxx_t *dev, uint32_t pos, uint8_t val, size_t len)
+{
+    int check = 0;
+    size_t wr = 0;
+    size_t chunk_len;
+
+    while (!check && (chunk_len = len - wr)) {
+        size_t max_chunk_len =
+            (AT24CXXX_PAGE_SIZE - (pos % AT24CXXX_PAGE_SIZE));
+        chunk_len = chunk_len < max_chunk_len ? chunk_len : max_chunk_len;
+
+        uint8_t polls = AT24CXXX_MAX_POLLS;
+        uint16_t reg = htons((uint16_t)pos);
+        while ((check =
+                    i2c_write_bytes(dev->params.i2c, dev->params.dev_addr, &reg,
+                                    AT24CXXX_ADDR_LEN,
+                                    I2C_NOSTOP)) == -ENXIO && --polls) {
+            xtimer_usleep(AT24CXXX_POLL_DELAY_MS * US_PER_MS);
+        }
+        DEBUG("_set::i2c_write_bytes check: %d; polls: %d\n", check, polls);
+        for (size_t i = 0; !check && i < chunk_len; i++) {
+            check = i2c_write_byte(dev->params.i2c, dev->params.dev_addr, val,
+                                   i + 1 <
+                                   chunk_len ? I2C_NOSTART |
+                                   I2C_NOSTOP : I2C_NOSTOP);
+            DEBUG("_set::i2c_write_byte check: %d\n", check);
+            if (!check) {
+                wr++;
+                pos++;
+            }
+        }
+    }
+    return wr > 0 ? (int)wr : check;
+}
+
+const at24cxxx_params_t *at24cxxx_get_default_dev_params(void)
+{
+    return &at24cxxx_params[0];
```

Comment:
Maybe it would be easier to return `const at24cxx_t *` and name the function `at24cxxx_get_default_dev()`, as every caller of this function directly does the cast.

Also, this would likely be easier to refactor. E.g. now every caller makes the assumptions, that the params can be casted to the device type. This assumption might not hold when the driver changes. If instead `get_default_dev()` directly returns the device descriptor, this assumption is only used in one place. So only one placed would need to be changed if this assumption no longer holds in future driver versions.

Reply:
Ok, I am going to change the return type.
If I wanted to inline the function, I had to put it into the public header file, where it could be visible for eepreg_wrapper.c. And if I put it into the public header file, the public header file also had to include the params header. But this is not the case for any other driver. Probably bacause any file that included the public header, would define a static params array.